### PR TITLE
Optimize needAck.

### PR DESCRIPTION
**Pull request by derekcollison**
Based off excellent insight and PR #3468 by @neilalexander.

This reuses pools we already had and also only loads the msg once for all consumers.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

**Review by kozlovic — APPROVED** [review]
LGTM. Let's see if @neilalexander can run that and see if that helps as it shown to help on the added test.

**Comment by derekcollison** [comment]
I added another optimization in case we are chasing lots of sequences below our first as well.

**Review by kozlovic — APPROVED** [review]
LGTM

**Review by neilalexander — APPROVED** [review]
This looks good — I've just tested it and it is no longer burning resources on `needAck` when creating the new ephemeral consumer after application startup.

File: server/consumer.go [hunk 1]
```diff
@@ -2328,17 +2328,22 @@ func (o *consumer) isFiltered() bool {
...
-	// Check first if we are filtered, and if so check if this is even applicable to us.
+	// Check if we are filtered, and if so check if this is even applicable to us.
 	if o.isFiltered() && o.mset != nil {
-		var svp StoreMsg
-		sm, err := o.mset.store.LoadMsg(sseq, &svp)
-		if err != nil || !o.isFilteredMatch(sm.subj) {
+		if subj == _EMPTY_ {
+			var svp StoreMsg
+			if _, err := o.mset.store.LoadMsg(sseq, &svp); err != nil {
+				return false
```
Comment (ashumkin) [inline server/consumer.go:2342]: I guess, `o.mu.RUnlock()` is missed here (before `return`)?

Comment (neilalexander) [reply server/consumer.go:2342]: Looks like it — https://github.com/nats-io/nats-server/pull/3555

**Pull request merged**